src/operation.rs

```diff
@@ -1656,69 +1700,81 @@ fn process_binary_operation(
+                                ts2.into_iter()
+                                    .map(|t| t.closurize(&mut env, env2.clone())),
+                            );
+
+                            Ok(Closure {
+                                body: RichTerm::new(Term::List(ts), pos_op_inh),
+                                env,
+                            })
+                        }
+                    } else {
+                        Err(EvalError::TypeError(
+                            String::from("List"),
+                            String::from("@, 2nd operand"),
+                            snd_pos,
+                            RichTerm {
+                                term: t2,
+                                pos: pos2,
+                            },
+                        ))
 
-                Ok(Closure {
-                    body: RichTerm::new(Term::List(ts), pos_op_inh),
-                    env,
-                })
-            }
-            (Term::List(_), t2) => Err(EvalError::TypeError(
-                String::from("List"),
-                String::from("@, 2nd operand"),
-                snd_pos,
-                RichTerm {
-                    term: Box::new(t2),
-                    pos: pos2,
-                },
-            )),
-            (t1, _) => Err(EvalError::TypeError(
-                String::from("List"),
-                String::from("@, 1st operand"),
-                fst_pos,
-                RichTerm {
-                    term: Box::new(t1),
-                    pos: pos1,
+                    }
                 },
-            )),
+            } else {
+                Err(EvalError::TypeError(
+                    String::from("List"),
+                    String::from("@, 1st operand"),
+                    fst_pos,
+                    RichTerm {
+                        term: t1,
+                        pos: pos1,
+                    },
+                ))
+            }
         },
-        BinaryOp::ListElemAt() => match (*t1, *t2) {
-            (Term::List(mut ts), Term::Num(n)) => {
-                let n_int = n as usize;
+        BinaryOp::ListElemAt() => match (&*t1, &*t2) {
+            (Term::List(ts), Term::Num(n)) => {
+                let n_int = *n as usize;
                 if n.fract() != 0.0 {
                     Err(EvalError::Other(format!("elemAt: expected the 2nd agument to be an integer, got the floating-point value {}", n), pos_op))
-                } else if n < 0.0 || n_int >= ts.len() {
+                } else if *n < 0.0 || n_int >= ts.len() {
                     Err(EvalError::Other(format!("elemAt: index out of bounds. Expected a value between 0 and {}, got {}", ts.len(), n), pos_op))
                 } else {
                     Ok(Closure {
-                        body: ts.swap_remove(n_int),
+                        // SAFETY: bound checking done in the previous step
+                        body: unsafe { ts.get_unchecked(n_int).clone() },
```

Comment:
I disagree, since we already made the checks before manually, why adding another one ?

Reply:
1. Using unsafe code is always a liability, and should have good motivations. Given the general state of performances in Nickel and the expected gain of performance (it's not like it's in a tight loop or something, then there is branch prediction, etc), I don't think eliding one bound check at this point is a very good motivation. We can always improve on performances later when the code is more mature.
2. IMHO it's ok to do it encapsulated in a datastructure or a macro because you have to change the internal code of said datastructure or macro to break the assumptions of unsafe blocks, in which that case you are usually (at least, you should) aware of the invariants you have to uphold. Here it is not really the case, as I myself wouldn't assume that I have to uphold some additional invariants in the middle of `process_binary_operation` or otherwise it could suddenly cause UB. Granted, there is a low probability of this code not doing the bound check anymore, and the unsafe block is quite close to the associated code, but it's still a very bad possibility.
3. It sounds like a job for the compiler, which is usually capable (well, LLVM actually) of doing this kind of things, especially in an `if-then-else` where `i >= len` is tested explicitly the branch before another. Out of curiosity I generated the assembly for `process_binary_operation` of the HEAD of this branch in release mode, and the same head with the unsafe removed and a `get().unwrap()` instead, and the output is *exactly* the same. If I mess with the bound (`n_int <= ts.len()+5`), then the output is different, with 11 more instructions in total (but much more changes, as it also shuffles some labels, registers, etc) for the safe version + buggy bound check. So in this specific case, right now, it seems it's  adding nothing.